nengo/base.py

```diff
@@ -50,18 +51,61 @@ class NengoObject(with_metaclass(NetworkMember, SupportDefaultsMixin)):
         The seed used for random number generation.
     """
 
+    _param_init_order = []  # Order in which parameters have to be initialized.
+                            # Missing parameters will be initialized last in an
+                            # undefined order. This needs to defined to allow
+                            # pickling and copying of Nengo objects if the
+                            # parameter initialization order matters.
+
     label = StringParam('label', default=None, optional=True)
     seed = IntParam('seed', default=None, optional=True)
 
     def __init__(self, label, seed):
+        super(NengoObject, self).__init__()
         self.label = label
         self.seed = seed
 
     def __getstate__(self):
-        raise NotImplementedError("Nengo objects do not support pickling")
+        state = self.__dict__.copy()
+        del state['_initialized']
+
+        for attr in get_params(self):
+            param = getattr(self.__class__, attr)
+            if self in param:
+                state[attr] = getattr(self, attr)
+
+        return state
 
     def __setstate__(self, state):
-        raise NotImplementedError("Nengo objects do not support pickling")
+        from nengo.network import Network
+
+        for attr in self._param_init_order:
+            setattr(self, attr, state.pop(attr))
+
+        for attr in get_params(self):
+            if attr in state:
+                setattr(self, attr, state.pop(attr))
+
+        for k, v in state.items():
+            setattr(self, k, v)
+
+        self._initialized = True
+        if len(Network.context) > 0:
+            warnings.warn(
+                "{obj} was not added to the network. When copying objects, "
+                "use the copy method on the object instead of Python's copy "
+                "module. When unpickling objects, they have to be added to "
+                "networks manually.".format(obj=self),
+                NotAddedToNetworkWarning)
+
+    def copy(self, add_to_container=True):
+        with warnings.catch_warnings():
```

Comment:
The initial copy is not added to a network and produces a warning, but the copy will be added to the network afterwards (if `add_to_container == True`) which means that the warning should not be shown (after all the copy is added to a network).

I assume you cannot do a `Network.add(copy(self))` either because the initial copied instance will still not be added to a network.